add CheckSigningTable config option

r-a-z-v-a-n opened the pull request:
When CheckSigningTable is set to no, the keys in KeyTable are no longer verified when config is loaded. 
This helps with large databases. This commit only adds support for USE_ODBX.

r-a-z-v-a-n committed:
add CheckSigningTable config option

When CheckSigningTable is set to no, the keys in KeyTable are no
longer verified when config is loaded. This helps with large
databases. This commit only adds support for USE_ODBX.

r-a-z-v-a-n committed:
conf_checksigningtable changed ==TRUE with !=FALSE

when comparing if (conf->conf_checksigningtable == TRUE)
and CheckSigningTable no in the config, no acts as TRUE.
Therefore, it is better to use != FALSE.

r-a-z-v-a-n commented:
Simple test results:
If the private key is bad:

```
Sep 08 19:32:56 bookworm01 opendkim[759]: OpenDKIM Filter v2.11.0 starting
Sep 08 19:59:13 bookworm01 opendkim[759]: 349634048D: SSL error:0680008E:asn1 encoding routines::not enough data
Sep 08 19:59:13 bookworm01 opendkim[759]: 349634048D: dkim_eom(): resource unavailable: d2i_PrivateKey_bio() failed

451 4.7.1 Service unavailable - try again later
```

If selector is NULL (should be prevented by sql not null option):

```
451 4.7.1 Service unavailable - try again later
Sep 08 20:06:12 bookworm01 postfix/cleanup[1546]: BCE424048F: milter-reject: END-OF-MESSAGE from localhost[::1]: 4.7.1 Service unavailable - try again later; from=<xxxx@linux>

Sep 08 20:06:12 bookworm01 opendkim[759]: KeyTable entry for '1' corrupt
Sep 08 20:06:12 bookworm01 opendkim[759]: BCE424048F: error loading key '1'
```

If key is NULL

```
Sep 08 20:09:16 bookworm01 opendkim[759]: KeyTable entry for '1' corrupt
Sep 08 20:09:16 bookworm01 opendkim[759]: 527A24048D: error loading key '1'

Sep 08 20:09:16 bookworm01 postfix/cleanup[1573]: 527A24048D: milter-reject: END-OF-MESSAGE from localhost[::1]: 4.7.1 Service unavailable - try again later; from=<xxx@linux>

```
in neither case did opendkim crash. Furthermore, you can solve a lot of these issues with db table constraints.

r-a-z-v-a-n commented:
this is continued from https://github.com/trusteddomainproject/OpenDKIM/pull/226

r-a-z-v-a-n commented:
also should there be a command line option for it ? example,
` opendkim/opendkim.c:15594:      /* process command line options */`
if yes, which letter may I use ?

futatuki commented on opendkim/opendkim.conf.5.in:
```diff
@@ -179,6 +179,11 @@ requires superuser access. A warning will be generated if
 
+.TP
+.I CheckSigningTable (Boolean)
+If set to yes, it walks the database on boot when it loads the config
```
"the database" is of course SigningTable, but I think it is good to write the table name explicitly. (Also in opendkim.conf sample)

r-a-z-v-a-n replied:
corrected in commit https://github.com/trusteddomainproject/OpenDKIM/pull/228/commits/906a8b48453b1e1367e56c84ac3d5e008080c823

r-a-z-v-a-n committed:
CheckSigningTable improve documentation

As advised by futatuki, specify SigningTable instead of the
"database" in the description.

futatuki commented:
> in neither case did opendkim crash. Furthermore, you can solve a lot of these issues with db table constraints.

I'll take a look later,  if it can avoid crash even if the key entry is corrupted. (But it is another issue)

futatuki commented:
> also should there be a command line option for it ? example, ` opendkim/opendkim.c:15594: /* process command line options */` if yes, which letter may I use ?

Currently 'C' is not in use, however it might be the time we shoud consider to use word options, although `getopt_long()` is not in POSIX...

r-a-z-v-a-n commented:
can we consider the long options issue as a separate issue?
one commit at a time so we dont get overloaded.

futatuki commented:
> can we consider the long options issue as a separate issue? one commit at a time so we dont get overloaded.

ya, sorry,  I think it is a separate issue.  It is not good that multiple issue in a PR.

futatuki commented:
This feature may useful even if ODBX feature is not used, so I think it is not need to restrict it on USE_ODBX is true.  No other things to say, it looks good to me, for the commits till [906a8b4](https://github.com/trusteddomainproject/OpenDKIM/pull/228/commits/906a8b48453b1e1367e56c84ac3d5e008080c823).

You can freely determine the command line option letter for this feature, if you implement it (perhaps no one complains about it).   However if I  implement a command line option for checking SiginigTable consistency and exit,  I'll select 'C' for it if it is still not in use :)

r-a-z-v-a-n committed:
CheckSigningTable option as -C argument

Allow the use of -C to disable CheckSigningTable (or set to no).

r-a-z-v-a-n committed:
CheckSigningTable make option always available

Allow disabling of CheckSigningTable for other databases such as
LDAP. This option disables the walking of SigningTable to look
for missing keys in KeyTable when the config gets loaded.

r-a-z-v-a-n commented:
see commit https://github.com/trusteddomainproject/OpenDKIM/pull/228/commits/898f6ec9410b1dd5cd4f192d86cce3a8891c21dc for args option -C
see commit https://github.com/trusteddomainproject/OpenDKIM/pull/228/commits/35f13b11770214e462ead7cb0551f6afa659b04e to allow for disabling CheckSigningTable for all databases.

r-a-z-v-a-n commented:
lhy is the develop branch so far ahead of master branch ? 
There are 127 pending commits, last one 2 years ago.
The master branch has not been updated in 6 years.
When is the next planned release ?

r-a-z-v-a-n committed:
CheckSigningTable use arg -g instead of -C

As requested by futatuki, I will use -g for CheckSigningTable
and reserve -C for future option to check the database tables.

r-a-z-v-a-n commented:
> You can freely determine the command line option letter for this feature, if you implement it (perhaps no one complains about it). However if I implement a command line option for checking SiginigTable consistency and exit, I'll select 'C' for it if it is still not in use :)

oops i just realized what you said.
I changed the option to -g in commit https://github.com/trusteddomainproject/OpenDKIM/pull/228/commits/ee40b42743857df4d23a2c602488d3deeb24ffb6

futatuki commented:
As you added a new command line option,  it should be described in opendkim(8) :)

> lhy is the develop branch so far ahead of master branch ?
> There are 127 pending commits, last one 2 years ago.
> The master branch has not been updated in 6 years.
> When is the next planned release ?

I think this project is not working any more.  So I made my own branch, maintain it, and I use here as a collection center of issues and proposals of the changes.

r-a-z-v-a-n commented:
Thank you @futatuki . I would like to commit these changes to your mirror as well.
I see your stable branch is 212 commits ahead : )

futatuki commented:
It would be great if you could add a description about new "-g" option to opendkim/opendkim.8.in?   I'd like to see your commit for it, not mine :)

r-a-z-v-a-n committed:
CheckSigningTable add -g arg to opendkim(8) man page

r-a-z-v-a-n commented:
see commit https://github.com/trusteddomainproject/OpenDKIM/pull/228/commits/3fc8cb7232e7bacc79b5031d72664c9d7c1c6f56 for -g in opendkim(8) man page.

futatuki commented on opendkim/opendkim.8.in:
```diff
@@ -282,6 +282,11 @@ Normally
+.I \-g
+Set CheckSigningTable to no.  This means that when the config is loaded,
+The SigningTable will not be checked for any missing keys in
+the KeyTable.
```
* A typo " the SiningTable"
* The mention to `CheckSigingTable` option is not so important for understanding what this commandline option does.  So it is good to move it after the last sentence of the paragraph, with reference information to opendkim.conf(5).

r-a-z-v-a-n replied:
I see a typo, but in your comment, not the code.
Anyway I re-wrote it based on your suggestion.
See commit https://github.com/trusteddomainproject/OpenDKIM/commit/3dabd5fc7f5151ad0eed067661aeb6c33b38d8ca

futatuki commented:
> see commit [3fc8cb7](https://github.com/trusteddomainproject/OpenDKIM/commit/3fc8cb7232e7bacc79b5031d72664c9d7c1c6f56) for -g in opendkim(8) man page.

Thank you!

Although I 'm not good at writing English, I've reviewed it.

r-a-z-v-a-n commented:
thank you so much for all your help futatuki!

r-a-z-v-a-n committed:
CheckSigningTable improve man page

r-a-z-v-a-n committed:
SigningTable improve man page

r-a-z-v-a-n reviewed (commented):
since we made it global, row doesnt always apply, example LDAP server.

futatuki commented:
Looks good. Thank you!

r-a-z-v-a-n committed:
remove extra line

futatuki committed:
opendkim: The -g option overrides CheckSigningTable setting on config file.

futatuki committed:
opendkim: Imprement -C option, just opposit of the -g option.

futatuki committed:
opendkim/opendkim.8.in: Update descriptions for -C and -g options.

r-a-z-v-a-n committed:
Merge pull request #2 from futatuki/check-signingtable-only

Allow overriding CheckSigningTable config option by command line options

r-a-z-v-a-n committed:
change -C to -G to match the opposite of -g

futatuki commented on opendkim/opendkim.8.in:
```diff
@@ -262,6 +263,14 @@ The value may include two different canonicalizations separated by a
 slash ("/") character, in which case the first will be applied to the
 headers and the second to the body.
 .TP
+.I \-G
```
Keep alphabetic order in the descriptions of command line options :)  This should be below of  `-g` option now.

r-a-z-v-a-n replied:
https://github.com/trusteddomainproject/OpenDKIM/pull/228/commits/ac1d2c6503109460a9c5ab9b07efc7942b82ac2e

futatuki replied:
New place of `-G` is still above of `-g`.  It seems the order is lowercase first. (Other changs looks good to me, thank you!)

r-a-z-v-a-n replied:
https://github.com/trusteddomainproject/OpenDKIM/pull/228/commits/dfa3dfef5d657a3452b6a5c56c282ec8060b8247

futatuki commented on opendkim/opendkim.conf.5.in:
```diff
@@ -179,6 +179,11 @@ requires superuser access. A warning will be generated if
 
+.TP
+.I CheckSigningTable (Boolean)
+If set to yes, it walks the SigningTable on boot when it loads the config
```
As I commented other place,  loading config (and the opportunity of the check) may occure other than on boot, by receiving SIGHUP signal.

r-a-z-v-a-n replied:
https://github.com/trusteddomainproject/OpenDKIM/pull/228/commits/ac1d2c6503109460a9c5ab9b07efc7942b82ac2e

r-a-z-v-a-n committed:
improve documentation and fix typo

r-a-z-v-a-n committed:
documentation place -G after -g

r-a-z-v-a-n committed:
reorder parameter list for brevity
